### ISSUE TITLE
fix(training): pre-staging hydrates the test split, with audio, even when test is disabled

### ISSUE THREAD

**Issue by ktinubu** [tinaudio/synth-setter#2872]
Part of #1741

## What happens

`prepare_data` materializes `test.lance` on every training run, including runs whose experiment sets `test: false`, and the test split is the only one projected **with the `audio` column**.

## Why

`src/synth_setter/data/vst_datamodule.py:29`

```python
_MATERIALIZE_SPLITS = ("train", "val", "test")
```

Hardcoded. `_derive_projection` (line 416) then builds a projection for all three:

```python
return {
    split: self._loader_columns(read_audio=split == predict_split)
    for split in _MATERIALIZE_SPLITS
}
```

`cfg.test` is never consulted, and `predict_split` defaults to the test shard, so `read_audio` is true for exactly that split. `materialize_splits` iterates `projection.items()`, so all three are staged.

## Impact

- Runs with `test: false` (the eval workflow owns test evaluation) pay a full test-split download before step 1 — and it is the largest projection of the three because of `audio`.
- The projection also feeds the `dataset_root` cache digest, so the hydration cache is keyed on columns the run never reads.
- Operationally this reads as a hang: train and val log `lance_materialize.done` within seconds of each other, then the run goes silent for a long time on `test.lance` with no further output.

## Reproduce

```bash
synth-setter-train experiment=surge/flow_sketch_440k \
  datamodule.download_dataset_row_limit=2048 \
  datamodule.batch_size=32 datamodule.num_workers=2 \
  trainer.max_steps=20 trainer.min_steps=20 \
  training.val_audio_probe=false logger=csv
```

Observed in the logs — note the extra column on the third line:

```
lance_materialize.start  columns=('param_array', 'mel_spec', 'sketch')  .../train.lance
lance_materialize.start  columns=('param_array', 'mel_spec', 'sketch')  .../val.lance
lance_materialize.start  columns=('param_array', 'mel_spec', 'sketch', 'audio')  .../test.lance
```

The composed config has `test: false` throughout.

## Expected

Stage only the splits the selected stages will read. A `fit`-only run should not hydrate `test.lance` at all; when it is needed, `audio` should only be projected if a prediction stage will actually read it.

## Not a Lance read-path problem

An isolated projected scan of the test split at the same row limit completes in the same time as train and val, so the cost is the volume the `audio` column adds, not a defect in the scanner.

**Comment by ktinubu** [comment]
### [W7] Residual scope after #2919, and why the obvious fix is not cheap

Recording the analysis so the next owner does not re-derive it. **Not claiming this** — no branch or worktree exists for it.

**What #2919 fixed.** `_staged_projection` narrows hydration to the splits the running Lightning stage reads, so a `test: false` fit run no longer downloads `test.lance`. The reported symptom — a long silent hang after train/val — is gone.

**What is left.** On `0f81e9c9d6`, `_predict_split(None, root)` returns `"test"` whenever `predict_file` is unset, and `datamodule/vst.yaml` ships `predict_file: null`. So `_derive_projection` puts `audio` on the test split for **every** run:

```python
split: self._loader_columns(read_audio=self.include_audio or split == predict_split)
```

Two consequences, both matching the issue's Impact list:

1. A fit-only run's `dataset_root` digest is keyed on a column no stage will read. `subset_dirname` hashes the *whole* projection dict, so this is not confined to the test entry.
2. A `mode: test` eval run still downloads `test.lance` **with** `audio`, though `trainer.test` does not read it. Only `mode: predict` does.

**Why it is not a one-liner.** The datamodule cannot know at `__init__` whether a predicting stage will run, and per-stage *column* narrowing is ruled out by the note `_staged_projection` already carries: the digest covers the whole read set, so narrowing columns would re-hydrate shared splits under a new digest and clash with an earlier stage's manifest. The signal that exists is `cfg.mode` in `eval.yaml` (`test` | `val` | `predict`); `train.yaml` has no `mode` and `trainer.predict` is only ever called from `cli/eval.py`. So the honest fix makes "a prediction stage will run" explicit at construction rather than implied by `predict_file: null`.

**The cost to weigh first.** Any change to the projection changes `subset_dirname`, and because the digest spans all splits, that invalidates **every existing cached subset** — forcing a full re-hydration of train and val, not just test. That is a real operational bill on top of a correctness win that is cache-hygiene, not a hang. It is worth deciding that trade deliberately rather than discovering it on the next training launch.

Suggest this is sized as its own Task against the #2353 digest contract rather than carried as the tail of a bug whose symptom is fixed.

**Comment by ktinubu** [comment]
### [Claude-W7] Still live on current `main`, but it collides head-on with open PR #3600 — sequencing note

Swept this as an unclaimed dataloader item and confirmed the body is still accurate at `0f81e9c9d6`: `src/synth_setter/data/vst_datamodule.py` still hardcodes

```python
_MATERIALIZE_SPLITS = ("train", "val", "test")
```

and `_derive_projection` still builds a projection for all three without consulting `cfg.test`, so a `fit`-only run still pays the full `audio`-bearing test projection before step 1.

**Not opening a PR yet, deliberately.** [PR #3600](https://github.com/tinaudio/synth-setter/pull/3600) (`Refs #2923`, another W7 session) is open against the same two files and the same three functions — `_MATERIALIZE_SPLITS`, `_derive_projection`, and `materialize_splits` in `lance_materialize.py`, plus both of their test modules. Changing the split set underneath it would conflict hard and would make that PR's own reds harder to read.

Right order is #3600 first, then this on top of the merged result — the fix here is *which* splits reach `materialize_splits`, which is cleanest to express once #3600 has settled *how* they are pinned to a snapshot set.

Recording so the next session does not start it blind and hit the same conflict. Happy for the #3600 owner to absorb this into that PR instead if they prefer; it is a small change to the same lines.

**Comment by ktinubu** [comment]
### [Claude-W7] Fixed by #2919 — hydration is stage-scoped, and the `test: false` run never touches `test.lance`

At `b6d9cb08b6` the hardcoded materialize set is still there, but it is no longer what `prepare_data` consults:

```python
_STAGE_MATERIALIZE_SPLITS = MappingProxyType({
    TrainerFn.FITTING: ("train", "val"),
    TrainerFn.VALIDATING: ("val",),
    TrainerFn.TESTING: ("test",),
})
```

`_MATERIALIZE_SPLITS` survives only as the fallback for a datamodule with no running stage, and as the key-validation set for `download_dataset_txids`. A fit run resolves `FITTING` → `("train", "val")`, so the third `lance_materialize.start` line from the filed log — the `audio`-bearing one — never happens.

**Both filed impacts are gone, and the second one falls out of the first**: `audio` is projected into the *test split's* projection (`read_audio = split == predict_split`), but that projection is only ever materialized when a testing or predicting stage is running. Nothing downloads a waveform column for a run that will not read it.

Nine stage-scoping tests now pin this directly, including `test_fit_stage_leaves_the_test_split_unhydrated` and `test_prepare_data_after_a_finished_fit_hydrates_the_remaining_splits` — that last one guards the subtle half, since a finished `Trainer` keeps its last `state.fn` and would otherwise scope a later call to a stage that is no longer running. `tests/data/test_lance_materialize_datamodule.py` → **34 passed**.

**One filed sub-point is still literally true and should not be "fixed".** The issue notes the `dataset_root` cache digest is computed from the full projection, including columns the run never reads — and it still is (`_resolve_dataset_root(configured_root, self.projection)`). That is now load-bearing rather than wasteful: the digest has to be stage-independent, or a fit run and a later test run against the same dataset would resolve different subset directories and the test stage could not extend what fit already hydrated. `test_later_test_run_extends_the_subset_directory_the_fit_run_hydrated` is the assertion of exactly that. Narrowing the digest per stage would re-break what #2919 fixed.

Closing. Part of #1741. The operational symptom described here (silent minutes after two fast `done` lines) is worth keeping in mind for any *other* hydration path, since it reads as a hang and is not one.
